### PR TITLE
feat: implement upgrade controller

### PR DESCRIPTION
**Pull request by Unix4ever**
Upgrade controller is using talos upgrade module underneath.
`UpgradeK8sTask` is the input for the controller, which contains:
- upgrade version to.
- upgrade version from.
- local context + cluster.

For each `UpgradeK8sTask` the controller creates a separate task which
then puts upgrade logs to `TaskLogs` and the task status to `TaskStatus`
object.

Updating the task will stop the ongoing upgrade and will start a new
one.
Destroying the task will stop the ongoing upgrade.

Logs and Statuses resources are not deleted to keep track of already
processed tasks.

Signed-off-by: Artem Chernyshev <artem.0xD2@gmail.com>

### CODE REVIEW

**Comment by andrewrynhard** [comment]
I wonder if we can automate `from`. We should have that info? I know this is how it works in `talosctl`, so more of a discussion than a request.

**Comment by smira** [comment]
> I wonder if we can automate `from`. We should have that info? I know this is how it works in `talosctl`, so more of a discussion than a request.

this one is a bit tricky, as `from` should be the lowest version of the component in the cluster

imagine upgrade went 50% through for example, and API server version is inconsistent

**Comment by Unix4ever** [comment]
> I wonder if we can automate from. We should have that info? I know this is how it works in talosctl, so more of a discussion than a request.

Yeah... I hope to get it automated somehow. So as an option it seems that we can have a preparation step to scan all the nodes and find the lowest version. What do you think?
There will be the last PR, that will add the UI and close gaps like this.

**Comment by andrewrynhard** [comment]
> > I wonder if we can automate from. We should have that info? I know this is how it works in talosctl, so more of a discussion than a request.
> 
> Yeah... I hope to get it automated somehow. So as an option it seems that we can have a preparation step to scan all the nodes and find the lowest version. What do you think?
> There will be the last PR, that will add the UI and close gaps like this.

That is what came to mind for me.

File: internal/backend/runtime/theila/controllers/upgrade.go [hunk 1]
```diff
@@ -25,49 +33,50 @@ const (
...
+	}
 }
 
 // Outputs implements controller.Controller interface.
 func (ctrl *UpgradeController) Outputs() []controller.Output {
 	return []controller.Output{
 		{
-			Type: resources.UpgradeK8sTaskType,
+			Type: resources.TaskStatusType,
+			Kind: controller.OutputShared,
```
Comment (smira) [inline internal/backend/runtime/theila/controllers/upgrade.go:69]:
```suggestion
			Kind: controller.OutputExclusive,
```

File: internal/backend/runtime/theila/controllers/upgrade.go [hunk 2]
```diff
@@ -82,8 +91,226 @@ func (ctrl *UpgradeController) Run(ctx context.Context, r controller.Runtime, lo
...
+		selfHosted, err := k8s.IsSelfHostedControlPlane(ctx, &state, endpoints[0])
+		if err != nil {
+			if err = update(0, err); err != nil {
+				logger.Error("failed to update status", zap.Error(err))
+
+				return
+			}
+		}
+
+		if selfHosted {
```
Comment (smira) [inline internal/backend/runtime/theila/controllers/upgrade.go:251]: remove me

File: internal/backend/runtime/theila/controllers/upgrade.go [hunk 3]
```diff
@@ -82,8 +91,226 @@ func (ctrl *UpgradeController) Run(ctx context.Context, r controller.Runtime, lo
...
+		endpoints := c.GetEndpoints()
+
+		selfHosted, err := k8s.IsSelfHostedControlPlane(ctx, &state, endpoints[0])
+		if err != nil {
+			if err = update(0, err); err != nil {
+				logger.Error("failed to update status", zap.Error(err))
+
+				return
+			}
+		}
```
Comment (smira) [inline internal/backend/runtime/theila/controllers/upgrade.go:249]: return here?

File: internal/backend/runtime/theila/controllers/upgrade.go [hunk 4]
```diff
@@ -82,8 +91,226 @@ func (ctrl *UpgradeController) Run(ctx context.Context, r controller.Runtime, lo
...
+
+	wg sync.WaitGroup
+}
+
+//nolint:gocognit
+func (t *upgradeTask) start(ctx context.Context, task *resources.UpgradeK8sTask, r controller.Runtime, logger *zap.Logger) error {
+	status := resources.NewTaskStatus(Namespace, task.Metadata().ID())
+
+	update := func(progress float32, err error) error {
+		return r.Modify(ctx, status, func(res resource.Resource) error {
```
Comment (smira) [inline internal/backend/runtime/theila/controllers/upgrade.go:157]: this is not supported: `controller.Runtime` is not goroutine-safe

**Comment by andrewrynhard** [comment]
I don't see the frontend changes (as in a button or something). Unless I am just missing it?

**Comment by Unix4ever** [comment]
frontend changes are coming in the next PR. It's going to be pretty huge on itself and I'm still working on it.

**Comment by andrewrynhard** [comment]
> So frontend changes are coming in the next PR. It's going to be pretty huge by itself and I'm still working on it.

Ah ok.

**Review by andrewrynhard — APPROVED** [review]

**Pull request merged**